File: src/main/java/AoC2021_01.java
import java.util.List;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2021_01 extends SolutionBase<List<Integer>, Long, Long> {

    private AoC2021_01(final boolean debug) {
        super(debug);
    }

    public static final AoC2021_01 create() {
        return new AoC2021_01(false);
    }

    public static final AoC2021_01 createDebug() {
        return new AoC2021_01(true);
    }

    @Override
    protected List<Integer> parseInput(final List<String> inputs) {
        return inputs.stream().map(Integer::valueOf).toList();
    }

    private long countIncreases(final List<Integer> depths, final int window) {
        return IntStream.range(window, depths.size())
                .filter(i -> depths.get(i) > depths.get(i - window))
                .count();
    }

    @Override
    public Long solvePart1(final List<Integer> depths) {
        return countIncreases(depths, 1);
    }

    @Override
    public Long solvePart2(final List<Integer> depths) {
        return countIncreases(depths, 3);
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "7"),
        @Sample(method = "part2", input = TEST, expected = "5"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2021_01.create().run();
    }

    private static final String TEST = """
            199
            200
            208
            210
            200
            207
            240
            269
            260
            263
            """;
}

File: src/main/cpp/2021/01/AoC2021_01.cpp
#include <assert.h>
#include <iostream>
#include <string>
#include <vector>
#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const vector<string> TEST = {
    "199",
    "200",
    "208",
    "210",
    "200",
    "207",
    "240",
    "269",
    "260",
    "263"
};

int countIncreases(const vector<string> &input, const int window) {
    int ans = 0;
    for (unsigned i = window; i < input.size(); i++) {
        if (stoi(input[i]) > stoi(input[i - window])) {
            ans++;
        }
    }
    return ans;
}

int part1(const vector<string> &input) {
    return countIncreases(input, 1);
}

int part2(const vector<string> &input) {
    return countIncreases(input, 3);
}

void samples() {
    assert(part1(TEST) == 7);
    assert(part2(TEST) == 5);
}

MAIN(2021, 1)
